File: internal/wails/schema.go
package wails

import (
	"context"

	"github.com/wzhejunqiu/data-nexus/internal/model"
	"github.com/wzhejunqiu/data-nexus/internal/service"
	"go.uber.org/zap"
)

type SchemaService struct {
	query *service.QueryService
	log   *zap.Logger
}

func NewSchemaService(query *service.QueryService, log *zap.Logger) *SchemaService {
	return &SchemaService{query: query, log: log}
}

func (s *SchemaService) ListTables(connectionID string) (*model.TableList, error) {
	return call(s.log, "SchemaService.ListTables", func() (*model.TableList, error) {
		return s.query.ListTables(context.Background(), connectionID)
	})
}

func (s *SchemaService) GetTableSchema(connectionID string, tableName string) (*model.TableSchema, error) {
	return call(s.log, "SchemaService.GetTableSchema", func() (*model.TableSchema, error) {
		return s.query.GetTableSchema(context.Background(), connectionID, tableName)
	})
}

func (s *SchemaService) GetTableProfile(connectionID string, tableName string) (*model.TableProfile, error) {
	return call(s.log, "SchemaService.GetTableProfile", func() (*model.TableProfile, error) {
		return s.query.GetTableProfile(context.Background(), connectionID, tableName)
	})
}

func (s *SchemaService) DetectFTSTable(connectionID string, tableName string) (*model.FTSInfo, error) {
	return call(s.log, "SchemaService.DetectFTSTable", func() (*model.FTSInfo, error) {
		return s.query.DetectFTSTable(context.Background(), connectionID, tableName)
	})
}

File: frontend/src/lib/api/schema.ts
import {
  DetectFTSTable,
  GetTableProfile,
  GetTableSchema,
  ListNamespaces,
  ListSchemas,
  ListTables,
} from '../../../wailsjs/go/wails/SchemaService'
import type {
  FTSInfo,
  ListTablesOptions,
  NamespaceList,
  SchemaList,
  TableList,
  TableProfile,
  TableSchema,
} from '../types'
import { mapWailsError } from './errors'

async function wrap<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn()
  } catch (err) {
    throw mapWailsError(err)
  }
}

export const schemaApi = {
  listNamespaces: (connectionId: string) =>
    wrap(() => ListNamespaces(connectionId) as Promise<NamespaceList>),
  listSchemas: (connectionId: string, database: string) =>
    wrap(() => ListSchemas(connectionId, database) as Promise<SchemaList>),
  listTables: (connectionId: string, opts: ListTablesOptions = {}) =>
    wrap(() => ListTables(connectionId, opts) as Promise<TableList>),
  getTableSchema: (connectionId: string, tableName: string) =>
    wrap(() => GetTableSchema(connectionId, tableName) as Promise<TableSchema>),
  getTableProfile: (connectionId: string, tableName: string) =>
    wrap(() => GetTableProfile(connectionId, tableName) as Promise<TableProfile>),
  detectFTS: (connectionId: string, tableName: string) =>
    wrap(() => DetectFTSTable(connectionId, tableName) as Promise<FTSInfo>),
}
